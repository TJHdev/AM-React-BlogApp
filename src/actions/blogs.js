import database from '../firebase/firebase.js'

// component calls action generator
// action generator returns object
// component dispatches object
// redux store changes

// component calls action generator
// action generator returns function
// component dispatches object (?)
// function runs (has the ability to dispatch other actions and do whatever it wants)

export const addBlog = (blog) => ({
  type: 'ADD_BLOG',
  blog: blog
})

export const startAddBlog = (blogData = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    const {
      title = '',
      body = '',
      createdAt = 0,
      lastEditedAt = 0
    } = blogData;
    
    const blog = { title, body, createdAt, lastEditedAt };

    return database.ref(`users/${uid}/blogs`).push(blog).then((ref) => {
      dispatch(addBlog({
        id: ref.key,
        ...blog
      }));
    });
  };
};

export const removeBlog = ({ id } = {}) => ({
  type: 'REMOVE_BLOG',
  id: id
});

export const startRemoveBlog = ({ id }) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/blogs/${id}`).set(null).then(() => {
      dispatch(removeBlog({ id }));
    });
  }
};

export const editBlog = (id, updates) => ({
  type: 'EDIT_BLOG',
  id: id,
  updates
});

export const startEditBlog = (id, updates = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/blogs/${id}`).update(updates).then(() => {
      dispatch(editBlog(id, updates));
    }); 
  };
};

export const setBlogs = (blogs) => ({
  type: 'SET_BLOGS',
  blogs: blogs
});

export const startSetBlogs = (blogsData = []) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/blogs`).once('value').then((snapshot) => {
      const blogs = [];
      snapshot.forEach((childSnapshot) => {
        blogs.push({
          id: childSnapshot.key,
          ...childSnapshot.val()
        })
      })
      dispatch(setBlogs(blogs));
    });
  }
};

export const startSetBlogsPublic = (blogsData = []) => {
  return (dispatch, getState) => {
    return database.ref('users').once('value').then((snapshot) => {
      const blogs = [];
      const val = snapshot.val();

      Object.keys(val).forEach((childSnapshot) => {
        Object.keys(val[childSnapshot].blogs).forEach((blog) => {
          blogs.push({
            id: blog,
            ...snapshot.val()[childSnapshot].blogs[blog]
          })
        })
      })
      dispatch(setBlogs(blogs));
    });
  }
};

