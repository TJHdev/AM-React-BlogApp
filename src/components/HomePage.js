import React from 'react';
import BlogListFilters from './BlogListFilters';
import BlogListPublic from './BlogListPublic';

const HomePage = () => (
  
  <div>
    <BlogListFilters />
    <BlogListPublic />
  </div>
);

export default HomePage;