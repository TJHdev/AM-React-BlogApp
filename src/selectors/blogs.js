import moment from 'moment';

// Get visible blogs

const getVisibleBlogs = (blogs, { text, sortBy }) => {
  return blogs.filter((blog) => {
    const textMatch = blog.title.toLowerCase().includes(text.toLowerCase());
    return textMatch;
  }).sort((a, b) => {
    if (sortBy === 'dateCreated') {
      return a.createdAt < b.createdAt ? 1 : -1;
    } else if (sortBy === 'dateEdited') {
      return a.lastEditedAt < b.lastEditedAt ? 1 : -1;
    } else if (sortBy === 'title') {
      return a.title < b.title ? 1 : -1;
    }
  });
};

export default getVisibleBlogs;