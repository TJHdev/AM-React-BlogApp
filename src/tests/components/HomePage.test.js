import { shallow } from 'enzyme';
import HomePage from '../../components/HomePage';
import blogs from '../fixtures/blogs';

test('should render HomePage correctly', () => {
  const wrapper = shallow(<HomePage />);
  expect(wrapper).toMatchSnapshot();
});