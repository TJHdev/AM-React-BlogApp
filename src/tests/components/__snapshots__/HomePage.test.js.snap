// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render HomePage correctly 1`] = `
<div>
  <Connect(BlogListFilters) />
  <Connect(BlogListPublic) />
</div>
`;
